File: thrift/test/py/TestOptionalRequired.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import unittest
import unittest2

from OptionalRequiredTest import *
from OptionalRequiredTest.ttypes import *

from thrift.protocol.TBinaryProtocol import *
from thrift.transport.TTransport import *

class TestOptionalRequired(unittest.TestCase):

    def testSetDefault(self):
        w = Tricky1()
        r = Tricky1()
        w.im_default = 10
        write_to_read(w, r)
        self.assertEquals(10, r.im_default)

    def testSetOptional(self):
        w = Tricky2()
        r = Tricky2()
        w.im_optional = 10
        write_to_read(w, r)
        self.assertEquals(10, r.im_optional)

    def testSetRequired(self):
        w = Tricky3()
        r = Tricky3()
        w.im_required = 10
        write_to_read(w, r)
        self.assertEquals(10, r.im_required)

    def testSetDefaultNull(self):
        w = Tricky1()
        r = Tricky1()
        w.im_default = None
        write_to_read(w, r)
        self.assertEquals(None, r.im_default)

    def testSetOptionalNull(self):
        w = Tricky2()
        r = Tricky2()
        w.im_optional = None
        write_to_read(w, r)
        self.assertEquals(None, r.im_optional)

    def testSetRequiredNull(self):
        w = Tricky3()
        r = Tricky3()
        w.im_required = None
        try:
            write_to_read(w, r)
            self.assertTrue(
                False,
                "Excepted exception because field is required"
            )
        except Exception:
            pass

    def testSetDefaultDontSet(self):
        w = Tricky1()
        r = Tricky1()
        write_to_read(w, r)
        self.assertEquals(None, r.im_default)

    def testSetOptionalDontSet(self):
        w = Tricky2()
        r = Tricky2()
        write_to_read(w, r)
        self.assertEquals(None, r.im_optional)

    def testSetRequiredDontSet(self):
        w = Tricky3()
        r = Tricky3()
        try:
            write_to_read(w, r)
            self.assertTrue(
                False,
                "Excepted exception because field is required"
            )
        except Exception:
            pass

    def testMixDefaultAndOptional(self):
        w = Tricky1()
        r = Tricky2()

        w.im_default = 0
        r.im_optional = 10

        write_to_read(w, r)
        self.assertEquals(0, r.im_optional)
        self.assertEquals(0, w.im_default)

        write_to_read(r, w)
        self.assertEquals(0, r.im_optional)
        self.assertEquals(0, w.im_default)

    def testMixDefaultAndRequired(self):
        w = Tricky1()
        r = Tricky3()

        w.im_default = 0
        r.im_required = 10

        write_to_read(w, r)
        self.assertEquals(0, r.im_required)
        self.assertEquals(0, w.im_default)

        write_to_read(r, w)
        self.assertEquals(0, r.im_required)
        self.assertEquals(0, w.im_default)


def write_to_read(write_struct, read_struct):
    write_buffer = TMemoryBuffer()
    write_protocol = TBinaryProtocol(write_buffer)
    write_struct.write(write_protocol)

    # The implementation of TMemoryBuffer is slightly different from C++
    # the read/write buffer is not shared, thus we have to create another
    # TMemoryBuffer
    read_buffer = TMemoryBuffer(write_buffer.getvalue())
    read_protocol = TBinaryProtocol(read_buffer)
    read_struct.read(read_protocol)

if __name__ == '__main__':
    unittest.main()

File: thrift/test/php/TestOptionalRequired.php
<?php
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

if (!isset($GEN_DIR)) {
  $GEN_DIR = 'gen-php';
}
if (!isset($MODE)) {
  $MODE = 'normal';
}

/** Set the Thrift root */
$GLOBALS['THRIFT_ROOT'] = '../../lib/php/src';

/** Include the Thrift required libraries */
require_once $GLOBALS['THRIFT_ROOT'].'/Thrift.php';
require_once $GLOBALS['THRIFT_ROOT'].'/protocol/TBinaryProtocol.php';
require_once $GLOBALS['THRIFT_ROOT'].'/transport/TMemoryBuffer.php';

require_once $GEN_DIR.'/OptionalRequiredTest_types.php';

/////////////////////////////////
// SETTING THE VALUE
/////////////////////////////////

// Default gets written (when set)
$w = new Tricky1();
$r = new Tricky1();

$w->im_default = 10;
write_to_read($w, $r);
assert_equals(10, $r->im_default);

// Optional gets written (when set)
$w = new Tricky2();
$r = new Tricky2();

$w->im_optional = 10;
write_to_read($w, $r);
assert_equals(10, $r->im_optional);

// Require gets written (when set)
$w = new Tricky3();
$r = new Tricky3();

$w->im_required = 10;
write_to_read($w, $r);
assert_equals(10, $r->im_required);

/////////////////////////////////
// SETTING THE VALUE TO NULL
/////////////////////////////////

// Default gets written (when set  TO NULL)
$w = new Tricky1();
$r = new Tricky1();

$w->im_default = NULL;
write_to_read($w, $r);
assert_equals(NULL, $r->im_default);

// Optional gets written (when set TO NULL)
$w = new Tricky2();
$r = new Tricky2();

$w->im_optional = NULL;
write_to_read($w, $r);
assert_true(!isset($r->im_optional));

// Require gets written (when set TO NULL)
$w = new Tricky3();
$r = new Tricky3();

$w->im_required = NULL;

try {
  write_to_read($w, $r);
  assert_true(false, "Expected exception because field is required");
} catch (Exception $ex) {  }
assert_true(!isset($r->im_required));

/////////////////////////////////
// WITHOUT SETTING THE VALUE
/////////////////////////////////

// Default gets written (when set)
$w = new Tricky1();
$r = new Tricky1();

write_to_read($w, $r);
assert_true(!isset($r->im_default));

// Optional gets written (when set)
$w = new Tricky2();
$r = new Tricky2();

write_to_read($w, $r);
assert_true(!isset($r->im_optional));

// Require gets written (when set)
$w = new Tricky3();
$r = new Tricky3();

try {
  write_to_read($w, $r);
  assert_true(false, "Expected exception because field is required");
} catch (Exception $ex) {  }
assert_true(!isset($r->im_required));

/////////////////////////////////
// Mixing Default & Optional & Required
/////////////////////////////////

// Default <-> Optional
$w = new Tricky1();
$r = new Tricky2();

$w->im_default = 0;
$r->im_optional = 10;

write_to_read($w, $r);
assert_equals(0, $r->im_optional);
assert_equals(0, $w->im_default);

write_to_read($r, $w);
assert_equals(0, $r->im_optional);
assert_equals(0, $w->im_default);

// Default <-> Required
$w = new Tricky1();
$r = new Tricky3();

$w->im_default = 0;
$r->im_required = 10;

write_to_read($w, $r);
assert_equals(0, $r->im_required);
assert_equals(0, $w->im_default);

write_to_read($r, $w);
assert_equals(0, $r->im_required);
assert_equals(0, $w->im_default);

echo "Tests passed!\n";

function write_to_read($write_struct, $read_struct) {
  $protocol = new TBinaryProtocol(new TMemoryBuffer());
  $write_struct->write($protocol);
  $read_struct->read($protocol);
}

function assert_true($bool) {
  if (!$bool) {
    throw new Exception("Assertion is false");
  }
}

function assert_equals($expected, $actual) {
  if ($expected != $actual) {
    throw new Exception("Variable was expected contain the value '$expected'" .
      " the actual value was '$actual'.");
  }
}

?>
